controllers/applicationvolumereplication_controller.go

```diff
@@ -195,658 +276,895 @@ func (r *ApplicationVolumeReplicationReconciler) Reconcile(ctx context.Context,
+func (a *AVRInstance) restorePVFromBackup(homeCluster string) error {
+	a.log.Info("Restoring PVs to new managed cluster", "name", homeCluster)
 
-	var homeCluster string
+	pvList, err := a.listPVsFromS3Store()
+	if err != nil {
+		return errorswrapper.Wrap(err, "failed to retrieve PVs from S3 store")
+	}
 
-	var peerCluster string
+	a.log.Info(fmt.Sprintf("Found %d PVs", len(pvList)))
 
-	switch {
-	case subStatuses[d1.Name] != nil:
-		homeCluster = d1.Name
-		peerCluster = d2.Name
-	case subStatuses[d2.Name] != nil:
-		homeCluster = d2.Name
-		peerCluster = d1.Name
-	default:
-		return empty, empty, fmt.Errorf("mismatch between placementRule %s decisions and subscription %s statuses",
-			placementRule.Name, subscription.Name)
+	if len(pvList) == 0 {
+		return nil
 	}
 
-	return homeCluster, peerCluster, nil
+	// Create manifestwork for all PVs for this AVR
+	return a.createOrUpdatePVsManifestWork(a.instance.Name, a.instance.Namespace, homeCluster, pvList)
 }
 
-func (a *AVRInstance) getManagedClustersUsingPlacementRule(
-	placementRule *plrv1.PlacementRule, maxClusterCount int) (map[string]*spokeClusterV1.ManagedCluster, error) {
-	const requiredClusterReplicas = 1
+func (a *AVRInstance) createOrUpdatePVsManifestWork(
+	name string, namespace string, homeClusterName string, pvList []corev1.PersistentVolume) error {
+	a.log.Info("Create manifest work for PVs", "AVR",
+		name, "cluster", homeClusterName, "PV count", len(pvList))
+
+	mwName := BuildManifestWorkName(name, namespace, MWTypePV)
 
-	clmap, err := utils.PlaceByGenericPlacmentFields(
-		a.reconciler.Client, placementRule.Spec.GenericPlacementFields, nil, placementRule)
+	manifestWork, err := a.generatePVManifestWork(mwName, homeClusterName, pvList)
 	if err != nil {
-		return nil, fmt.Errorf("failed to get cluster map for placement %s error: %w", placementRule.Name, err)
+		return err
 	}
 
-	if placementRule.Spec.ClusterReplicas != nil && *placementRule.Spec.ClusterReplicas != requiredClusterReplicas {
-		return nil, fmt.Errorf("PlacementRule %s Required cluster replicas %d != %d",
-			placementRule.Name, requiredClusterReplicas, *placementRule.Spec.ClusterReplicas)
-	}
+	return a.createOrUpdateManifestWork(manifestWork, homeClusterName)
+}
 
-	err = a.filterClusters(placementRule, clmap)
-	if err != nil {
-		return nil, fmt.Errorf("failed to filter clusters. Cluster len %d, error (%w)", len(clmap), err)
-	}
+func (a *AVRInstance) processVRGManifestWork(homeCluster string) error {
+	a.log.Info("Processing VRG ManifestWork", "cluster", homeCluster)
 
-	if len(clmap) != maxClusterCount {
-		return nil, fmt.Errorf("PlacementRule %s should have made %d decisions. Found %d",
-			placementRule.Name, maxClusterCount, len(clmap))
+	if err := a.createOrUpdateVRGRolesManifestWork(homeCluster); err != nil {
+		a.log.Error(err, "failed to create or update VolumeReplicationGroup Roles manifest")
+
+		return err
 	}
 
-	return clmap, nil
-}
+	if err := a.createOrUpdateVRGManifestWork(
```

Comment:
(doubt) Will the namespace exist on the cluster this is being targeted to? When does the namespace get created?

Reply:
The namespace must exist.  That's the user's responsibility.  The order that ACM shows is to create namespace, channel, placementrule, subscription, then application.  We will inject ourselves in between the placementrule and the subscription.  If the ManifestWork creation fails for any reason, we will report the error and retry at later time.

Reply:
Ack! As we update userPlRule and then create VRG, ACM should eventually create the ns for VRG creation to succeed and we will fail and reconcile if we race to VRG creation first.